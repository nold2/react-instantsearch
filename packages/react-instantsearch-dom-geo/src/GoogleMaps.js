import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { LatLngPropType, BoundingBoxPropType } from './propTypes';

export const GOOGLE_MAPS_CONTEXT = '__ais_geo_search__google_maps__';

class GoogleMaps extends Component {
  static propTypes = {
    cx: PropTypes.func.isRequired,
    google: PropTypes.object.isRequired,
    initialZoom: PropTypes.number.isRequired,
    initialPosition: LatLngPropType.isRequired,
    mapOptions: PropTypes.object.isRequired,
    refine: PropTypes.func.isRequired,
    position: LatLngPropType,
    boundingBox: BoundingBoxPropType,
    children: PropTypes.node,
  };

  static childContextTypes = {
    [GOOGLE_MAPS_CONTEXT]: PropTypes.shape({
      google: PropTypes.object,
      instance: PropTypes.object,
    }),
  };

  isUserInteraction = true;
  isPendingRefine = false;
  listeners = [];

  state = {
    isMapReady: false,
  };

  getChildContext() {
    const { google } = this.props;

    return {
      [GOOGLE_MAPS_CONTEXT]: {
        instance: this.instance,
        google,
      },
    };
  }

  componentDidMount() {
    const { google, mapOptions } = this.props;

    this.instance = new google.maps.Map(this.element, {
      mapTypeControl: false,
      fullscreenControl: false,
      streetViewControl: false,
      clickableIcons: false,
      zoomControlOptions: {
        position: google.maps.ControlPosition.LEFT_TOP,
      },
      ...mapOptions,
    });

    google.maps.event.addListenerOnce(
      this.instance,
      'idle',
      this.setupListenersWhenMapIsReady
    );
  }

  componentDidUpdate() {
    const {
      google,
      initialZoom,
      initialPosition,
      boundingBox,
      position,
    } = this.props;

    if (boundingBox) {
      this.lockUserInteration(() => {
        this.instance.fitBounds(
          new google.maps.LatLngBounds(
            boundingBox.southWest,
            boundingBox.northEast
          ),
          0
        );
      });
    }

    if (!boundingBox) {
      const initialMapPosition = position || initialPosition;

      this.lockUserInteration(() => {
        this.instance.setZoom(initialZoom);
        this.instance.setCenter(initialMapPosition);
      });
    }
  }

  componentWillUnmount() {
    this.listeners.forEach(listener => {
      listener.remove();
    });

    this.listeners = [];
  }

  setupListenersWhenMapIsReady = () => {
    const { refine } = this.props;

    this.setState(() => ({
      isMapReady: true,
    }));

    const onChange = () => {
      if (this.isUserInteraction) {
        this.isPendingRefine = true;
      }
    };

    this.listeners.push(this.instance.addListener('center_changed', onChange));
    this.listeners.push(this.instance.addListener('zoom_changed', onChange));
    this.listeners.push(this.instance.addListener('dragstart', onChange));

    this.listeners.push(
      this.instance.addListener('idle', () => {
        if (this.isUserInteraction && this.isPendingRefine) {
          this.isPendingRefine = false;

          const bounds = this.instance.getBounds();

          refine({
            northEast: bounds.getNorthEast().toJSON(),
            southWest: bounds.getSouthWest().toJSON(),
          });
        }
      })
    );
  };

  lockUserInteration(functionThatAlterTheMapPosition) {
    this.isUserInteraction = false;
    functionThatAlterTheMapPosition();
    this.isUserInteraction = true;
  }

  render() {
    const { cx, children } = this.props;
    const { isMapReady } = this.state;

    return (
      <div className={cx('')}>
        <div ref={c => (this.element = c)} className={cx('map')} />
        {isMapReady && children}
      </div>
    );
  }
}

export default GoogleMaps;
