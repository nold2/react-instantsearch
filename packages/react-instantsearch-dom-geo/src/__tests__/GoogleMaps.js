import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  createFakeGoogleReference,
  createFakeMapInstance,
} from '../../test/fakeGoogleMaps';
import GoogleMaps from '../GoogleMaps';

Enzyme.configure({ adapter: new Adapter() });

describe('GoogleMaps', () => {
  const defaultProps = {
    google: createFakeGoogleReference(),
    initialZoom: 1,
    initialPosition: {
      lat: 0,
      lng: 0,
    },
    position: null,
    boundingBox: null,
    cx: x => `geo ${x}`.trim(),
  };

  it('expect render correctly without the map rendered', () => {
    const props = {
      ...defaultProps,
    };

    const wrapper = shallow(
      <GoogleMaps {...props}>
        <div>This is the children</div>
      </GoogleMaps>,
      {
        disableLifecycleMethods: true,
      }
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state()).toEqual({
      isMapAlreadyRender: false,
    });
  });

  it('expect render correctly with the map rendered', () => {
    const props = {
      ...defaultProps,
    };

    const wrapper = shallow(
      <GoogleMaps {...props}>
        <div>This is the children</div>
      </GoogleMaps>,
      {
        disableLifecycleMethods: true,
      }
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state()).toEqual({
      isMapAlreadyRender: false,
    });

    // Simulate didMount
    wrapper.instance().componentDidMount();

    // Trigger the update
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state()).toEqual({
      isMapAlreadyRender: true,
    });
  });

  describe('creation', () => {
    it('expect to create the GoogleMaps on didMount with the default options', () => {
      const fakeGoogle = createFakeGoogleReference();

      const props = {
        ...defaultProps,
        google: fakeGoogle,
      };

      mount(<GoogleMaps {...props} />);

      expect(fakeGoogle.maps.Map).toHaveBeenCalledTimes(1);
      expect(fakeGoogle.maps.Map).toHaveBeenCalledWith(
        expect.any(HTMLDivElement),
        {
          mapTypeControl: false,
          fullscreenControl: false,
          streetViewControl: false,
          clickableIcons: false,
          zoomControlOptions: {
            position: 'left:top',
          },
        }
      );
    });

    it('expect to create the GoogleMaps on didMount witht the given options', () => {
      const fakeGoogle = createFakeGoogleReference();

      const props = {
        ...defaultProps,
        google: fakeGoogle,
        mapOptions: {
          streetViewControl: true,
          otherOptionToPass: false,
        },
      };

      mount(<GoogleMaps {...props} />);

      expect(fakeGoogle.maps.Map).toHaveBeenCalledTimes(1);
      expect(fakeGoogle.maps.Map).toHaveBeenCalledWith(
        expect.any(HTMLDivElement),
        {
          mapTypeControl: false,
          fullscreenControl: false,
          streetViewControl: true,
          clickableIcons: false,
          otherOptionToPass: false,
          zoomControlOptions: {
            position: 'left:top',
          },
        }
      );
    });
  });

  describe('update', () => {
    it('expect to call fitBounds on didUpdate when boundingBox is provided', () => {
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      google.maps.LatLngBounds.mockImplementation((sw, ne) => ({
        northEast: ne,
        southWest: sw,
      }));

      const props = {
        ...defaultProps,
        google,
      };

      const wrapper = shallow(
        <GoogleMaps {...props}>
          <div>This is the children</div>
        </GoogleMaps>
      );

      expect(mapInstance.fitBounds).toHaveBeenCalledTimes(0);

      expect(mapInstance.setZoom).toHaveBeenCalledTimes(0);
      expect(mapInstance.setCenter).toHaveBeenCalledTimes(0);

      wrapper.setProps({
        boundingBox: {
          northEast: {
            lat: 10,
            lng: 10,
          },
          southWest: {
            lat: 14,
            lng: 14,
          },
        },
      });

      expect(mapInstance.fitBounds).toHaveBeenCalledTimes(1);
      expect(mapInstance.fitBounds).toHaveBeenCalledWith({
        northEast: {
          lat: 10,
          lng: 10,
        },
        southWest: {
          lat: 14,
          lng: 14,
        },
      });

      expect(mapInstance.setZoom).toHaveBeenCalledTimes(0);
      expect(mapInstance.setCenter).toHaveBeenCalledTimes(0);
    });

    it('expect to call setCenter & setZoom when boundingBox is not provided', () => {
      const mapInstance = createFakeMapInstance();
      const google = createFakeGoogleReference({
        mapInstance,
      });

      google.maps.LatLngBounds.mockImplementation((sw, ne) => ({
        northEast: ne,
        southWest: sw,
      }));

      const props = {
        ...defaultProps,
        google,
      };

      const wrapper = shallow(
        <GoogleMaps {...props}>
          <div>This is the children</div>
        </GoogleMaps>
      );

      expect(mapInstance.fitBounds).toHaveBeenCalledTimes(0);

      expect(mapInstance.setZoom).toHaveBeenCalledTimes(0);
      expect(mapInstance.setCenter).toHaveBeenCalledTimes(0);

      wrapper.setProps();

      expect(mapInstance.fitBounds).toHaveBeenCalledTimes(0);

      expect(mapInstance.setZoom).toHaveBeenCalledWith(1);
      expect(mapInstance.setCenter).toHaveBeenCalledWith({
        lat: 0,
        lng: 0,
      });
    });
  });
});
