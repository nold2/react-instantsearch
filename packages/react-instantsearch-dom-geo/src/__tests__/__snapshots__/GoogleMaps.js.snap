// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMaps expect render correctly with the map rendered 1`] = `
<div
  className="ais-GeoSearch"
>
  <div
    className="ais-GeoSearch-map"
  />
</div>
`;

exports[`GoogleMaps expect render correctly with the map rendered 2`] = `
<div
  className="ais-GeoSearch"
>
  <div
    className="ais-GeoSearch-map"
  />
  <div>
    This is the children
  </div>
</div>
`;

exports[`GoogleMaps expect render correctly without the map rendered 1`] = `
<div
  className="ais-GeoSearch"
>
  <div
    className="ais-GeoSearch-map"
  />
</div>
`;
